### ISSUE TITLE
Prediction using regression_forest

### ISSUE THREAD

**Issue by tw10** [grf-labs/grf#330]
**Description of the issue**
I use regression_forest to estimate the propensity score in my data. While predict(forest) gives predictions that has a correlation of 0.68 with Y, predict(forest, training data), which should give similar result, gives predictions that has basically zero correlation with Y. Not sure why this happens?

The similar issue occurs with the example given in the R Documentation as well.

Also, in regression_forest.R, there is a function called regression_train. Where can I find this function?

Any help or clarification is appreciated.

**Steps to reproduce**
n = 50; p = 10
X = matrix(rnorm(n*p), n, p)
Y = X[,1] * rnorm(n)
r.forest = regression_forest(X, Y)
r.pred = predict(r.forest)$predictions
r.pred1 = predict(r.forest,X)$predictions
cor(r.pred,r.pred1)
cor(r.pred,Y)
cor(r.pred1,Y)

**GRF version**
0.10.1

**Comment by halflearned** [comment]
Hi @tw10, thanks for the issue!

This is a problem of how `predict` parses the second argument. To ensure that you get the same output, please use:

```
r.pred1 = predict(r.forest, newx=X)$predictions
```

This is an issue I've encountered before. I wonder if there's a way to deal with it. I'll investigate and maybe continue the discussion here.

**Comment by susanathey** [comment]
@halflearned meant to say we should use newdata, not newx, sorry!

**Comment by swager** [comment]
The difference between those two lines is that if you don't give `newdata` as an argument to `predict`, then we provide out-of-bag predictions on the training set (i.e., we only consider trees not using the i-th training example when predicting at Xi). On the other hand, if you set `newdata=X`, then we don't use out-of-bag estimation (and treat `X` as though it were a new test set).

In this simulation example, note that `E[Y|X=x] = 0` regardless of `x`, so we'd expect mu.hat(Xi) to be roughly unrelated to Yi for a well calibrated predictor (as is the case when we use out-of-bag predictions). If we don't use out-of-bag prediction, then mu.hat(Xi) overfits to Yi, and we observe a positive correlation.

As a clarification, note that the following two lines do exactly the same thing (i.e., non out-of-bag prediction on `X`)

```R
r.pred1 = predict(r.forest,X)$predictions
r.pred1 = predict(r.forest, newdata=X)$predictions
```

I'm a little confused about what the phenomenon behind original question is, as here the correlation between r.pred1 and Y is systematically large and positive.

**Comment by tw10** [comment]
Thank you all for the helpful replies!

> In this simulation example, note that `E[Y|X=x] = 0` regardless of `x`, so we'd expect mu.hat(Xi) to be roughly unrelated to Yi for a well calibrated predictor (as is the case when we use out-of-bag predictions). If we don't use out-of-bag prediction, then mu.hat(Xi) overfits to Yi, and we observe a positive correlation.

This clears my confusion!

> I'm a little confused about what the phenomenon behind original question is, as here the correlation between r.pred1 and Y is systematically large and positive.

As for the strange result I got using my data in the first post, I trained the model using only a few selected columns of X, while predicting using the whole data frame. Silly mistake!
